src/mds/SessionMap.h

```diff
@@ -255,28 +262,61 @@ class Session : public RefCountedObject {
 
 class MDS;
 
-class SessionMap {
+/**
+ * Encapsulate the serialized state associated with SessionMap.  Allows
+ * encode/decode outside of live MDS instance.
+ */
+class SessionMapStore {
+protected:
+    mds_rank_t rank;
 public:
-  MDS *mds;
-private:
   ceph::unordered_map<entity_name_t, Session*> session_map;
+  version_t version;
+
+  virtual void encode(bufferlist& bl) const;
+  virtual void decode(bufferlist::iterator& blp);
+  void dump(Formatter *f) const;
+
+  Session* get_or_add_session(const entity_inst_t& i) {
+    Session *s;
+    if (session_map.count(i.name)) {
+      s = session_map[i.name];
+    } else {
+      s = session_map[i.name] = new Session;
+      s->info.inst = i;
+      s->last_cap_renew = ceph_clock_now(g_ceph_context);
+    }
+
+    return s;
+  }
+
+  static void generate_test_instances(list<SessionMapStore*>& ls);
+
+  void reset_state()
+  {
+    session_map.clear();
+  }
+
+  SessionMapStore() : rank(MDS_RANK_NONE), version(0) {}
```

Comment:
So the SessionMap encoding is badly done, as the decode reads sessions until the end of the stream (while !p.end()) so adding any new fields to the end isn't possible without a breaking change.  The good news is that we're going to completely change the encoding anyway to avoid huge OSD ops on updates, so that problem will go away soon :-)

In the MDS class, we definitely should be waiting until a rank is acquired before construction SessionMap.  However, that's a broader issue for all the rank-specific subsystems, and having a NULL sessionmap member lying around without a getter protecting it feels dangerous to me.  I want to split up MDS into MDSDaemon and MDSRank at some stage, where the latter part only gets initialized when the daemon acquires a rank.

Anyway -- what I suggest is that we make the setting of rank more explicit via an init() method, and then assert that we're initialized when adding sessions (and also encoding/decoding for good measure).  That should make this class sufficiently safe and clear, while deferring the deeper issues of encoding format and object lifetime to when we can address them properly.